### PR TITLE
[PostRector] Skip re-import name on callable name node

### LINKED ISSUE #3383: PHP Fatal Error: Cannot redeclare Functional\*
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.7.26                                                          |
| Installed as   | composer dependency                                             |

## Minimal PHP Code Causing Issue

There seems to be some kind of problem in a project using [lstrojny/functional-php](https://github.com/lstrojny/functional-php). What I've seen so far is that seems to crash when a file imports a function that was already imported in another file. Related issues #3145. I've created an example repository with failing code.

https://github.com/theUniC/rectorphp-report-issue

The CLI I've used is

	php vendor/bin/rector process --dry-run -vvvvv

## Expected Behaviour

To finish the process without fatal errors.

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #3383

### CODE REVIEW

**Comment by TomasVotruba** [comment]
This will need a failing test case first, so we're sure it's fixed

**Comment by samsonasik** [comment]
yes, it works functional `use function` exists like in #3383, but it cause failure on not exists with auto-import in `Rector\Renaming\Tests\Rector\Name\RenameClassRector\FunctionAutoImportNamesParameterTest::test`

**Comment by samsonasik** [comment]
we may need to replicate `Rector\CodingStyle\ClassNameImport\ClassNameImportSkipper::isFoundInUse()`:

https://github.com/rectorphp/rector/blob/94a4bb77778599c7de7550872a3cc4f68843f50c/rules/coding-style/src/ClassNameImport/ClassNameImportSkipper.php#L53-L68

**Comment by samsonasik** [comment]
still have 1 error:

```
1) Rector\CodingStyle\Tests\Rector\Namespace_\ImportFullyQualifiedNamesRector\ImportFullyQualifiedNamesRectorTest::test with data set #14 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules/coding-style/tests/Rector/Namespace_/ImportFullyQualifiedNamesRector/Fixture/bootstrap_names.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 <?php
 
 namespace {
-    use function SomeNamespace\gc_disable;
     function run_me_never()
     {
         // silent deprecations, since we test them
@@ @@
         error_reporting(E_ALL ^ E_DEPRECATED);
 
         // performance boost
-        gc_disable();
+        \SomeNamespace\gc_disable();
     }
 }

/home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:486
/home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:308
/home/runner/work/rector/rector/rules/coding-style/tests/Rector/Namespace_/ImportFullyQualifiedNamesRector/ImportFullyQualifiedNamesRectorTest.php:27
```

**Comment by samsonasik** [comment]
finally fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: tests/Issues/Issue3383/DoNotReImportFunction/config/some_config.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+use Rector\Core\Configuration\Option;
+use Rector\Set\ValueObject\SetList;
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+
+return static function (ContainerConfigurator $containerConfigurator): void {
+    $parameters = $containerConfigurator->parameters();
+
+    $parameters->set(Option::SETS, [SetList::DEAD_CODE]);
+    $parameters->set(Option::AUTO_IMPORT_NAMES, true);
+};
```
Comment (TomasVotruba) [inline tests/Issues/Issue3383/DoNotReImportFunction/config/some_config.php:14]: :+1:

**Comment by bendavies** [comment]
Still getting the error from #3383 unfortunately. I'll try to debug it now i can see your fix.

**Comment by samsonasik** [comment]
@bendavies I tried with `rector-functional-php` repository and it working ok:

Screen Shot 2021-01-08 at 16 25 03

You can use `dev-master` as in the screenshot.

**Comment by bendavies** [comment]
Thanks @samsonasik .

Definitely doesn't work on another project of mine, using the master version of rector.
I'll look into it.

**Comment by samsonasik** [comment]
Just in case it by cache, you can run with `--clear-cache`:

```bash
vendor/bin/rector --clear-cache process
```

**Comment by TomasVotruba** [comment]
@bendavies To save both yours and ours time, a GitHub repository with failing GitHub Action would be helpful.
That way, we'll see exactly where the problem is. And we all would know if the problem is solved or not.

**Comment by bendavies** [comment]
hi @samsonasik i've replicated the bug again here:
https://github.com/bendavies/rector-functional-php

@TomasVotruba i have tried to reproduce in github actions but strangely it didn't replicated.
It errors locally to me.
I'd appreciate if someone could try?

I'll continue to try and work out why it errors locally for me.

**Comment by bendavies** [comment]
I've also added a stack trace on that repo readme

**Comment by bendavies** [comment]
well, adding
```php
        if (function_exists($classLike)) {
            return true;
        }
```

to https://github.com/rectorphp/rector/blob/96bd6e2464dc34c1b56913757b45a5ee407b84ad/packages/node-type-resolver/src/ClassExistenceStaticHelper.php#L9 fixes the issue for me.

**Comment by bendavies** [comment]
I think the problem night be here?
https://github.com/rectorphp/rector/blob/96bd6e2464dc34c1b56913757b45a5ee407b84ad/packages/post-rector/src/Rector/NameImportingPostRector.php#L118

`Functional\each` will do `function_exists('each')` which is always true, since `each` is a built in ph p function

**Comment by samsonasik** [comment]
@bendavies I did a lot of simulation in that part :), that's used in rules, that's why that check was needed, could you provide a patch and add a fixture to https://github.com/rectorphp/rector/tree/master/tests/Issues/Issue3383/DoNotReImportFunction/Fixture for reproduce it? thank you.

**Comment by samsonasik** [comment]
@bendavies I cloned your repo, re-try run `vendor/bin/rector process` and it working ok:

Screen Shot 2021-01-08 at 20 12 28

That maybe related with cache somewhere, please try clear cache composer as well.

**Comment by samsonasik** [comment]
@bendavies you may need to remove `vendor` first too, as that something-something cache sometime there, remove composer.lock, composer clearcache, remove re-try composer install, and re-run `vendor/bin/rector --clear-cache process`

That's it, I can't reproduce it by using your repo. If you can provide reproduceable fixture, and provide a fix, that will be awesome ;)

**Comment by bendavies** [comment]
@samsonasik thanks for trying. i'll try to get a reproduced in this repo.

**Pull request merged**